src/include/simeng/config/Expectations.hh

```diff
@@ -0,0 +1,403 @@
+    static_assert(is_expected_type<T>::value &&
+                  "[SimEng:ModelConfig] Unexpected type given to "
+                  "ExpectationNode::getByType");
+
+    // Value existence check
+    if (variant.valueless_by_exception()) {
+      std::cerr << "[SimEng:ModelConfig] No value in passed variant within "
+                   "ExpectationNode with key "
+                << nodeKey_ << std::endl;
+      exit(1);
+    }
+    // Value type check
+    if (!std::holds_alternative<T>(variant)) {
+      std::cerr << "[SimEng:ModelConfig] A value of given type not held in "
+                   "variant within ExpectationNode with key "
+                << nodeKey_ << ". Variant holds a "
+                << typeToString(variant.index())
+                << " and the expected type of this node is "
+                << typeToString(static_cast<size_t>(type_)) << "." << std::endl;
+      exit(1);
+    }
+    return std::get<T>(variant);
+  }
+
+  /** A utility function for converting the type held in DataTypeVariant or the
+   * value of type_ into a string via an index. */
+  std::string typeToString(size_t index) const {
+    switch (index) {
+      case static_cast<size_t>(ExpectedType::Bool):
+        return "bool";
+      case static_cast<size_t>(ExpectedType::Double):
+        return "double";
+      case static_cast<size_t>(ExpectedType::Float):
+        return "float";
+      case static_cast<size_t>(ExpectedType::Integer):
+        return "integer";
+      case static_cast<size_t>(ExpectedType::String):
+        return "string";
+      case static_cast<size_t>(ExpectedType::UInteger):
+        return "unsigned integer";
+    }
+    return "unknown";
+  }
+
+  /** Setter function to set the default value for this node's associated config
+   * option. */
+  void setDefaultValue(DataTypeVariant var) { defaultValue_ = var; }
+
+  /** Setter function to set the expected bounds for this node's associated
+   * config option. */
+  template <typename T>
+  void setValueBounds(T lower, T upper) {
+    definedBounds_ = true;
+    expectedBounds_.first = lower;
+    expectedBounds_.second = upper;
+  }
+
+  /** Setter function to set the expected set of values for this node's
+   * associated config option. */
+  template <typename T>
+  void setValueSet(std::vector<T> set) {
+    definedSet_ = true;
+    for (const T s : set) {
+      DataTypeVariant dtv = s;
+      expectedSet_.push_back(dtv);
+    }
+  }
+
+  /** Add a child node to the vector of children within this node. */
+  void addChild(ExpectationNode chld) {
+    // Ensure that if the new child is wild, one does not already exist in this
+    // instance's children
+    if (chld.getKey() == wildcard) {
+      if (hasWild_) {
+        std::cerr << "[SimEng:ModelConfig] Attempted to add multiple wild "
+                     "nodes to the same ExpectationNode instance of key "
+                  << nodeKey_ << std::endl;
+        exit(1);
```

Comment:
The way you want this is only somewhat possible. Due to the wildcard stuff, you'd never be able to print "Ports::1::Portname", only "Ports::*::Portname" which doesn't match your intent as far as I can tell. This is because that "1" doesn't exist in the expectation tree, only the config tree passed in

Reply:
The feasibility of doing it with the wildcard stuff issue is okay. We'd have to build a cumulative string which is passed down to each child when the parent adds it to its list of children.

Reply:
I'm not 100% sure on your intent here as the original question is flawed (i.e. "Ports::1::Portname" doesn't exist) so I don't if you meant to be applying this printout to the expectationTree checking or configTree validation

Reply:
That was a bad example on my part. What I meant was printing the cumulative string to this node, i.e. `[Parent][Some-Child][Some-grandchild]` or in the case of a wildcard `[Parent][*][some-grandchild]`